doc/asynchronous_fifo/asynchronous_fifo.rst

```diff
@@ -0,0 +1,463 @@
+time the resampled value was taken. This interpolated time stamp is passed
+on to the Asynchronous FIFO, noting that only the last sample value was
+calculated.
+
+The consuming side just picks a sample of the queue and provides a timestamp::
+
+        int now = (void) hwtimer_get_time(tmr);
+        asynchronous_fifo_consume(a, &output_data, now);
+
+Ideally, the timestamp is taken when, for example, the previous sample was
+put on the wire and is therefore as close as possible to the word clock of
+the outgoing signal. Note that if the timestamp is related to the previous
+sample, then a constant of one sample-time should be added to the time
+stamp to make sure that the queue is exactly in the middle and not one off
+the middle.
+
+Using the Asynchronous FIFO with a PLL
+++++++++++++++++++++++++++++++++++++++
+
+
+Internal workings of the Asynchronous FIFO
+------------------------------------------
+
+Internally, the Asynchronous FIFO measures the phase difference between
+producer and the consumer. It does so by storing the time samples of all
+consumed samples, and comparing the sample consumed N/2 samples in the past
+with the time sample 
+TWhat we are going to measure is the phase-difference between samples going
+into the FIFO and samples going out of the FIFO. In an ideal world, sample X
+sample enters the FIFO at exactly the same time as sample X-N/2 leaving the
+FIFO. N/2 is the ideal fill-level of the FIFO that is N items long. For
+example, for a FIFO of 10 elements the ideal fill level is 5. If the
+output is running slightly too fast then sample X will enter the FIFO just
+after X-N/2 leaves the FIFO; if the output is running slightly too slow than
+sample X will enter the FIFO just before X-N/2 leaves the FIFO. We will be
+using the time between sample X entering and sample X-N/2 leaving the FIFO as
+the guidance for our rate matching algorithm; this is the phase difference.
+
+The phase difference is notionally a continuous value (a time stamp)
+in practice we measure it with the reference clock which has a 10 ns
+granularity. Unlike traditional phase differences that are measured in
+radians and where the maximum phase difference is *+/- pi* , we measure the
+phase difference as a time, and we allow the phase to be off by more than
+half a sample. If the queue is overflowing badly (say there is 9 items in
+it rather than 5), then the phase difference will account for the 4 extra
+items in the FIFO, and the fractional item, somewhere between 4 and 4.999...
+
+Note that the phase difference itself is an integral value; it is the
+number of samples since the beginning of time that the ASRC is out by. The goal of
+the rate converter is to make the phase difference stable (ie, it does not
+move between subsequent samples), and zero (ie, the FIFO is exactly mid
+level). Hence, we can see the differential of the phase error as a
+proportional error, and the phase error itself as an integral error.
+
```

Comment:
The Kp and Ki terms have been pre-calculated to cover all normal cases and do not need to be modified.